UI modifications to match Android, handle UNVERIFIED syncs, and more!

scottnonnenberg opened the pull request:
Verification notifications:
- We were showing identity key notifications too often (when we didn't have an identity key before, for example)
- We were showing verification notifications too often (when we got them from contact syncs)
- Because contact sync can send us `UNVERIFIED` keys, a number of changes were needed in the `Conversation` model, our `onVerified` handler, and in `processVerifiedMessage`.

A number of UI updates to match Android: 
- More informative strings which talk about verifying a safety number, and not the contact themselves
- A notification that a group message did not reach all recipients
- When outgoing identity key errors happen (because a getProfile didn't happen beforehand) we have a nicer experience on the message detail screen. A View button like Android, which opens an informative screen, with three options: 'Show Safety Number', 'Send Anyway', and 'Cancel'
- To ensure that the user sees the errors in large groups, the contacts with errors are sorted to the top. Otherwise it's alphabetical.

Miscellaneous:
- Our 'send anyway' button in the confirmation dialog didn't properly break through the five-second warning period after a key change; fixed.
- Fix code coverage instrumentation so that we get coverage for protocol layer (startup takes longer)
- Fix a few errors in tests, both under code coverage and generally
- Turned on the `laxbreak` jshint rule, because leading `||` and `&&` makes complex `if` subclauses easier to read.

liliakai commented on js/models/conversations.js:
```diff
@@ -142,8 +142,7 @@
                 || (keychange && verified === VERIFIED)) {
 
-                var local = !options.viaSyncMessage && !options.viaContactSync;
-                this.addVerifiedChange(this.id, verified === VERIFIED, {local: local});
```
Looks like this ensures that changes in state caused by contact syncs are not annotated as "from another device"... Android has no equivalent case for this choice since it never receives contact syncs. What's your reasoning here? Certainly if we got the change from a sync message, it didn't occur on this device.

scottnonnenberg replied:
A later commit changes it so `local` depends entirely on `viaSyncMessage`

liliakai commented on _locales/en/messages.json:
```diff
@@ -158,10 +158,13 @@
+      "message": "Mark as verified"
+    },
+    "unverify": {
+      "message": "Clear verification"
```
I think "Mark as not verified" is better than "Clear verification" for consistency with other copy like "You marked your safety number with $name$ as unverified." One verb to rule them all.

scottnonnenberg replied:
That's fair. But I will submit that 'Clear verification' is what iOS uses, and it's the only other app with a button for this action. Android has that switch with a simple label 'Verified.'

liliakai commented on test/keychange_listener_test.js:
```diff
@@ -1,5 +1,6 @@
 describe('KeyChangeListener', function() {
   var phoneNumberWithKeyChange = '+13016886524';  // nsa
+  var address = new libsignal.SignalProtocolAddress(phoneNumberWithKeyChange, 1);
```
What crash? This isn't used anywhere else in the file as of this rev. I added the same line in session-archive though, so maybe this is somehow an artifact of that?

scottnonnenberg replied:
Ah, I see what happened. This branch was previously on top of the `session-archive` branch, and when I rebased straight back on top of `combined`, this line was orphaned. It was replacing a line which caused a crash because it was referencing a nonexistent `identifier` variable, if I recall correctly.

scottnonnenberg commented on js/views/message_detail_view.js:
```diff
@@ -98,14 +132,21 @@
+                });
+
+                _.sortBy(contacts, function(c) {
+                    var prefix = this.grouped[c.id] ? '0' : '1';
```
Adding a prefix to the title strings gives us both alphabetical sorting and sorting by errors. It will be like this:

```
[
'0Alice', // with errors
'0Bob', // with errors
'1Alice (GVoice)',
'1John'
]
```

liliakai replied:
I see. First time through, I totally missed that there was a concatenation happening in the return statement.

scottnonnenberg commented on js/signal_protocol_store.js:
```diff
@@ -331,10 +331,10 @@
                         }));
                     }
-                    Promise.all(promises).then(resolve);
+                    Promise.all(promises).then(resolve, resolve);
```
I can move it to `reject`, no problem. It is being called from a non-test location: `removeIdentityKey` in this same file.

liliakai replied:
Yeah, but removeIdentityKey is only called in tests at this point. Now that we can overwrite keys when they changed there is never a need to remove them outside tests.

scottnonnenberg replied:
And I guess it's still necessary for test setup/teardown? If not, we should remove the dead code...

liliakai commented on js/views/conversation_view.js:
```diff
@@ -620,7 +627,10 @@
+                conversation: this.model,
+                // we pass these in to allow nested panels
+                listenBack: this.listenBack.bind(this),
+                resetPanel: this.resetPanel.bind(this)
```
I'm still not crazy about these triple(+?)ly nested panels. It makes it a little harder to keep track of the fact that you're not just making a decision about a certain contact, but a certain message. But then again, hopefully our new key change policies ensure that users reach this UX even more rarely than they used to.

scottnonnenberg replied:
Do you think that moving to a popup dialog like Android would make things easier to understand?

Alternatively, one way to improve things here would be to replicate the message that failed to send somewhere on the three-button warning page to make it clear what 'Send anyway' applies to.

liliakai replied:
I'm not sure what the best solution is. In the commits where I introduced the detail-expand UI, I included something to the effect of "One less modal", which seemed like a win at the time. Let's not block on it now though since it should be extremely edge casey now with all the profile fetching we do.

scottnonnenberg commented on js/views/conversation_view.js:
```diff
@@ -202,6 +202,13 @@
+
+        markAllAsApproved: function(untrusted) {
+            return Promise.all(untrusted.map(function(contact) {
+                return contact.setApproved();
```
Yup! Clearly the right name was `setApproved()`!

scottnonnenberg commented on js/models/conversations.js:
```diff
@@ -202,6 +202,14 @@
             }.bind(this)));
         }
     },
+    setTrusted: function() {
```
Sounds good. Updated.

scottnonnenberg commented:
Once things are looking good, I'll merge the fix-up commits into the body of the PR. I thought it would be easier to add new commits addressing your feedback so the re-review would be easier. :0)
